File: springboot-java-25/03-springboot-controller/src/main/java/com/ganatan/starter/continents/ContinentController.java
package com.ganatan.starter.continents;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicInteger;

import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.server.ResponseStatusException;

@RestController
@RequestMapping("/continents")
public class ContinentController {

  private final AtomicInteger idCounter = new AtomicInteger(0);
  private final List<Continent> continentList = new ArrayList<>();

  public record Continent(int id, String name) {
  }

  public ContinentController() {
    init("Africa", "America", "Asia", "Europe", "Oceania", "Antarctica");
  }

  @GetMapping({"", "/"})
  public List<Continent> getAllContinents() {
    return continentList;
  }

  @GetMapping("/{id}")
  public Continent getContinentById(@PathVariable int id) {
    return requireById(id);
  }

  @PostMapping({"", "/"})
  @ResponseStatus(HttpStatus.CREATED)
  public Continent createContinent(@RequestBody Continent body) {
    return create(body.name());
  }

  @PutMapping("/{id}")
  public Continent updateContinent(@PathVariable int id, @RequestBody Continent body) {
    Continent existing = requireById(id);
    int index = continentList.indexOf(existing);
    Continent updated = new Continent(id, body.name());
    continentList.set(index, updated);
    return updated;
  }

  @DeleteMapping("/{id}")
  @ResponseStatus(HttpStatus.NO_CONTENT)
  public void deleteContinent(@PathVariable int id) {
    Continent existing = requireById(id);
    continentList.remove(existing);
  }

  private void init(String... names) {
    for (String name : names) {
      create(name);
    }
  }

  private Continent create(String name) {
    int id = idCounter.incrementAndGet();
    Continent created = new Continent(id, name);
    continentList.add(created);
    return created;
  }

  private Continent requireById(int id) {
    return findById(id).orElseThrow(() ->
      new ResponseStatusException(HttpStatus.NOT_FOUND, "CONTINENT_NOT_FOUND"));
  }

  private Optional<Continent> findById(int id) {
    for (Continent c : continentList) {
      if (c.id() == id) {
        return Optional.of(c);
      }
    }
    return Optional.empty();
  }
}

File: nestjs/02-nest-controller/src/continent.controller.ts
import {
  Body, Controller, Delete, Get, HttpCode, HttpStatus, NotFoundException, Param,
  ParseIntPipe, Post, Put,
} from '@nestjs/common';

type Continent = {
  id: number;
  name: string;
};

@Controller('continents')
export class ContinentController {
  private idCounter = 0;
  private continentList: Continent[] = [];

  constructor() {
    ['Africa', 'America', 'Asia', 'Europe', 'Oceania', 'Antarctica']
      .forEach(name => this.createContinent({ name }));
  }

  @Get()
  getAllContinents(): Continent[] {
    return this.continentList;
  }

  @Get(':id')
  getContinentById(@Param('id', ParseIntPipe) id: number): Continent {
    const continent = this.findContinentById(id);
    if (!continent) {
      throw new NotFoundException();
    }
    return continent;
  }

  @Post()
  @HttpCode(HttpStatus.CREATED)
  createContinent(@Body() body: { name: string }): Continent {
    const continent: Continent = {
      id: ++this.idCounter,
      name: body.name,
    };
    this.continentList.push(continent);
    return continent;
  }

  @Put(':id')
  updateContinent(
    @Param('id', ParseIntPipe) id: number,
    @Body() body: { name: string },
  ): Continent {
    const index = this.findIndexById(id);
    if (index === -1) {
      throw new NotFoundException();
    }
    const updated: Continent = { id, name: body.name };
    this.continentList[index] = updated;
    return updated;
  }

  @Delete(':id')
  @HttpCode(HttpStatus.NO_CONTENT)
  deleteContinent(@Param('id', ParseIntPipe) id: number): void {
    const index = this.findIndexById(id);
    if (index === -1) {
      throw new NotFoundException();
    }
    this.continentList.splice(index, 1);
  }

  private findContinentById(id: number): Continent | undefined {
    return this.continentList.find(c => c.id === id);
  }

  private findIndexById(id: number): number {
    return this.continentList.findIndex(c => c.id === id);
  }
}